### PR TITLE
Ignore unguarded-availability for unit test

### PR DESCRIPTION
**Pull request by cyanglaz**
Ignore ungarded-availability for unit test

fixes https://github.com/flutter/flutter/issues/128958

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by cyanglaz** [comment]
When this lands successfully and pass tests post commit, I will follow up with a patch to updating the variable directly, then revert hard code value here.

**Comment by cyanglaz** [comment]
@hellohuanlin 
`@available(iOS 17) ` returns true when we set the min deploy version to ios 17 even when the simulator is ios 16, I'm changing it to use `NSProcessInfo` for a more accurate version check on device.

**Comment by hellohuanlin** [comment]
> @available(iOS 17)  returns true when we set the min deploy version to ios 17 even when the simulator is ios 16

I'm confused. If min deployment version is 17, how can it run on ios 16? Or am i missing something?

**Comment by cyanglaz** [comment]
> I'm confused. If min deployment version is 17, how can it run on ios 16? Or am i missing something?

Sorry I wasn't clear in the PR title. It is updating the c flag`-mios-simulator-version-min` to 17.0 when building the testing library. The min deployment target is still the one set in the xcproj.

**Review by hellohuanlin — APPROVED** [review]
as discussed offline, reminder to 
1. fix the missing `@available(iOS 17.0, *)`
2. comment why we need this change

**Comment by cyanglaz** [comment]
After offline discussion with @stuartmorgan, we think it is better to ignore the warning for now instead of updating the flag.

**Review by hellohuanlin — APPROVED** [review]
LGTM

File: shell/platform/darwin/ios/BUILD.gn [hunk 1]
```diff
@@ -227,6 +227,7 @@ source_set("ios_test_flutter_mrc") {
     "-F$platform_frameworks_path",
     "-mios-simulator-version-min=$ios_testing_deployment_target",
   ]
+  cflags_objcc = [ "-Wno-unguarded-availability-new" ]
```
Comment (stuartmorgan) [inline shell/platform/darwin/ios/BUILD.gn:230]: Please add a comment explaining why this is here, and what the conditions are for getting rid of it (i.e., the XCTest header bug being fixed).

File: shell/platform/darwin/ios/BUILD.gn [hunk 2]
```diff
@@ -271,6 +272,7 @@ shared_library("ios_test_flutter") {
     "-fobjc-arc",
     "-mios-simulator-version-min=$ios_testing_deployment_target",
   ]
+  cflags_objcc = [ "-Wno-unguarded-availability-new" ]
```
Comment (stuartmorgan) [inline shell/platform/darwin/ios/BUILD.gn:275]: Same.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: shell/platform/darwin/ios/BUILD.gn [hunk 3]
```diff
@@ -227,6 +227,12 @@ source_set("ios_test_flutter_mrc") {
     "-F$platform_frameworks_path",
     "-mios-simulator-version-min=$ios_testing_deployment_target",
   ]
+
+  # An XCode15-beta has a bug where iOS 17 API usage is not guarded.
```
Comment (stuartmorgan) [inline shell/platform/darwin/ios/BUILD.gn:231]: Nit: Xcode 15 beta

File: shell/platform/darwin/ios/BUILD.gn [hunk 4]
```diff
@@ -271,6 +277,12 @@ shared_library("ios_test_flutter") {
     "-fobjc-arc",
     "-mios-simulator-version-min=$ios_testing_deployment_target",
   ]
+
+  # An XCode15-beta has a bug where iOS 17 API usage is not guarded.
```
Comment (stuartmorgan) [inline shell/platform/darwin/ios/BUILD.gn:281]: Same.

**Review by stuartmorgan — APPROVED** [review]
LGTM with nit. Thanks!

**Pull request merged**